device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -505,17 +584,48 @@ public void onAuthenticationFailed(TransportException transportException)
     {
         this.savedException = transportException;
         releaseLatch(authenticationSessionOpenedLatch);
-        releaseLatch(deviceSessionsOpenedLatch);
+        releaseDeviceSessionLatchs();
+    }
+
+    @Override
+    public void onSessionClosedUnexpectedly(ErrorCondition errorCondition, String deviceId)
+    {
+        TransportException savedException = AmqpsExceptionTranslator.convertFromAmqpException(errorCondition);
+
+        if (this.deviceClientConfigs.size() > 1)
+        {
+            // When multiplexing, don't kill the connection just because a session dropped.
+            log.error("Amqp session closed unexpectedly. notifying the transport layer to start reconnection logic...", this.savedException);
+            scheduleDeviceSessionReconnection(savedException, deviceId);
+        }
+        else
+        {
+            // When not multiplexing, reconnection logic will just spin up the whole connection again.
+            this.savedException = savedException;
+            log.error("Amqp session closed unexpectedly. Closing this connection...", this.savedException);
+            this.connection.close();
+        }
     }
 
     @Override
-    public void onSessionClosedUnexpectedly(ErrorCondition errorCondition)
+    public void onCBSSessionClosedUnexpectedly(ErrorCondition errorCondition)
```

Comment:
What is CBS besides a TV network?

In .NET we'd case this as Cbs. All caps for Java?

Reply:
"CBS" is an acronym for "Claims Based Security"

Reply:
We don't typically lowercase the tail end of an acronym in Java, no

Reply:
Can we add some code comments for things like this? Spell out CBS and maybe provide a link to what it is somewhere.

Reply:
There is no formal documentation on this CBS link that I could link to, but I can add some comments to clarify what CBS is